### PR TITLE
All: add Stellar operation manageBuyOffer

### PR DESCRIPTION
**Pull request by MisterTicot**
I think I covered everything but tests: tests for similar operations manageOffer/createPassiveOffer were missing too.

I did not test the implementation as I don't know about how to proceed. Please let me know if anything is missing or wrong.

### CODE REVIEW

**Comment by tsusanka** [comment]
Please run `pipenv run make gen` to generate protobuf messages.

-----

Yes, you are right that some tests for Stellar operations are missing (see #2). These device tests however are a great start to test your implementation. Have a look on the [first test](https://github.com/trezor/trezor-firmware/blob/4c640bcb3e3594ec0d4e369f9834c42791c02355/tests/device_tests/test_msg_stellar_sign_transaction.py#L79) for example and try to write a test for manageBuyOffer, it should be very similar. There is a comment on the top of the file explaining how to get a fixture from Stellar's labrotory.

**Comment by MisterTicot** [comment]
@tsusanka Can you take it from here? I feel like I'm wasting huge amount of time trying to figure out something that you could solve in a dozen of minutes. Meanwhile, nobody is busy writing tests & updates for my own softwares.

**Comment by tsusanka** [comment]
What is the difference between StellarManageBuyOfferOp and the StellarManageOfferOp we already had?

**Comment by matejcik** [comment]
> What is the difference between StellarManageBuyOfferOp and the StellarManageOfferOp we already had? Is it completely unrelated op? Or do we use StellarManageOfferOp for "sell"?

Correct, "manage offer" is old name for "manage sell offer". We might want to rename "manage offer" to avoid confusion, however, it would break existing Connect users.
OTOH there probably aren't any, so  ¯\\\_(ツ)\_/¯.

Another thing i notice is that the messages are identical in terms of fields. The same code could handle them, and we could even fold them into a single message with a `BUY/SELL` enum flag. Leaving aside the fact that this is what Stellar devs should have done in the first place, this is where the atypical way Stellar is implemented is biting us in the behind.

I mean
we *can* make three distinct protobuf messages with the same fields (`CreatePassiveOffer` *also* has identical fields, except offer id is missing - but that's optional anyway so again ¯\\\_(ツ)\_/¯)
but we should reuse as much existing ManageOffer implementation as possible.
and that is like 98 % of it.

File: common/protob/messages-stellar.proto [hunk 1]
```diff
@@ -123,6 +124,21 @@ message StellarManageOfferOp {
...
+/**
+ * Request: ask device to confirm this operation type
+ * @next StellarTxOpRequest
+ * @next StellarSignedTx
+ */
+message StellarManageBuyOfferOp {
+    optional string source_account = 1;             // (optional) source account address
+    optional StellarAssetType selling_asset = 2;
+    optional StellarAssetType buying_asset = 3;
+    optional sint64 buyAmount = 4;
```
Comment (matejcik) [inline common/protob/messages-stellar.proto]: let's keep the name `amount`

Comment (MisterTicot) [reply common/protob/messages-stellar.proto]: The reason I used buyAmount is because it's what Stellar use (I guess it should be `buy_amount` by the way?). I know it's not consistent.

Do you want to switch to amount anyway?

**Comment by MisterTicot** [comment]
> Correct, "manage offer" is old name for "manage sell offer". We might want to rename "manage offer" to avoid confusion, however, it would break existing Connect users.
OTOH there probably aren't any, so ¯\_(ツ)_/¯.

stellaport.io is one of the major DEX front-end and has integrated Trezor since nearly one year. Naturally, they use the manage(Sell)Offer operation - so I'd assume there are people using it.

**Comment by MisterTicot** [comment]
Note: the latest HEAD passed the checks locally, but when I build T firmware & try to sign createPassiveOffer/manageSellOffer/manageBuyOffer ops I get `Firmware error`. I've tried to rebase on trezor-firmware master but got the same error.

**Comment by MisterTicot** [comment]
**Pull request rebased**

Example of transactions leading to firmware error: [manageSellOffer](https://test.cosmic.link/?manageSellOffer&amount=10&selling=GILS:gils*cosmic.plus&network=test&price=1).

**Comment by MisterTicot** [comment]
Anybody here???

**Comment by MisterTicot** [comment]
I spent time to play your game: learn how to use your development environment & write some missing code in Trezor support for Stellar. I had to figure out by myself how to make it fits your system.

Yet that PR is being ignored since a month. This is totally unrespectful. :angry:

**Comment by matejcik** [comment]
hello, you still haven't addressed this comment https://github.com/trezor/trezor-firmware/pull/627#issuecomment-543112157:

> Another thing i notice is that the messages are identical in terms of fields. The same code could handle them
(...)
> but we should reuse as much existing ManageOffer implementation as possible.
> and that is like 98 % of it.

I am sorry that i don't have more time to spend on 3rd party contributions in order to provide more detailed guidance. However, this is a low-priority issue for us.

**Comment by MisterTicot** [comment]
That issue was already present in your code. As you noted, there was already a duplicate for `passiveSellOffer` - I simply followed your ways.

I think there's a difference between me writing an update for your firmware & me refactoring your code base to fix a weakness in it. I'm not proficient at C & I never proposed to go that far.

I agree that this would be better but I fear that if you really want it then you'll have to write it by yourself.

File: common/protob/messages-stellar.proto [hunk 2]
```diff
@@ -123,6 +124,21 @@ message StellarManageOfferOp {
...
+/**
+ * Request: ask device to confirm this operation type
+ * @next StellarTxOpRequest
+ * @next StellarSignedTx
+ */
+message StellarManageBuyOfferOp {
+    optional string source_account = 1;             // (optional) source account address
+    optional StellarAssetType selling_asset = 2;
+    optional StellarAssetType buying_asset = 3;
+    optional sint64 buy_amount = 4;
```
Comment (matejcik) [inline common/protob/messages-stellar.proto]: please rename to `amount`

File: core/src/apps/stellar/operations/layout.py [hunk 3]
```diff
@@ -110,6 +123,19 @@
     await confirm_asset_issuer(ctx, op.buying_asset)
 
 
+async def _confirm_buy_offer(ctx, title, op):
```
Comment (matejcik) [inline core/src/apps/stellar/operations/layout.py]: please extend and reuse `_confirm_offer` instead

File: core/src/apps/stellar/operations/serialize.py [hunk 4]
```diff
@@ -68,6 +69,15 @@ def write_manage_offer_op(w, msg: StellarManageOfferOp):
     writers.write_uint64(w, msg.offer_id)
 
 
+def write_manage_buy_offer_op(w, msg: StellarManageBuyOfferOp):
```
Comment (matejcik) [inline core/src/apps/stellar/operations/serialize.py]: please reuse `write_manage_offer_op` instead

**Comment by matejcik** [comment]
I pointed out some places where reuse should happen. You're unfortunately right about the legacy code, where we can't use duck typing to save on copy-pasting.

File: common/protob/messages-stellar.proto (hunk 2 above)
Comment (MisterTicot) [reply common/protob/messages-stellar.proto]: You already asked for it the last time then closed it after I replied:

> The reason I used buyAmount is because it's what Stellar use (I guess it should be buy_amount by the way?). I know it's not consistent.
> 
> Do you want to switch to amount anyway?

I think it's better to stay consistent with the underlying implementation regardless of our opinion about it. (I asked them to change it into `amount` at a lower level)

File: core/src/apps/stellar/operations/layout.py (hunk 3 above)
Comment (MisterTicot) [reply core/src/apps/stellar/operations/layout.py]: Looking into it, `_confirm_offer` and `_confirm_buy_offer` have few lines in common (4 out of 8). Whatever way you put it, the relation between buying asset & selling asset is inversed - so is the price - so I see no way to rewrite `_confirm_offer` without crippling code readability.

File: core/src/apps/stellar/operations/serialize.py (hunk 4 above)
Comment (MisterTicot) [reply core/src/apps/stellar/operations/serialize.py]: I'm waiting for your response about `amount` so I can decide how to rewrite that one.

**Comment by MisterTicot** [comment]
@matejcik I'm waiting for your answer.

File: common/protob/messages-stellar.proto (hunk 2 above)
Comment (matejcik) [reply common/protob/messages-stellar.proto]: Right, sorry about that. Yes, please switch to `amount` anyway.

File: core/src/apps/stellar/operations/layout.py (hunk 3 above)
Comment (matejcik) [reply core/src/apps/stellar/operations/layout.py]: point taken, let's keep it

**Comment by matejcik** [comment]
i'm just now noticing that your code in core serializes **selling** then **buying** asset, but the legacy code does the opposite, i.e. **buying** then **selling**. which is correct?

i assumed that the code in core is correct, and the serialization for buy-offer and sell-offer is identical.

if it isn't, then please fix what is a bug in `write_manage_buy_offer_op`, and keep it separate from `write_manage_offer_op`. (I'm leaving up to you whether to rename `buy_amount` to `amount`; I intended the rename to simplify our implementation, but if it turns out that the serialization is different, we can keep the name consistent.

If it is identical, then please fix the order in legacy code.

The code in `python/src/trezorlib/stellar.py` might also need the fix? because the assets there are also ordered as **selling** then **buying**.

---

This is the final decision and I'm ready to merge once this is clarified.
I.e., if the serialization is *identical* between buy and sell, then please change field name to `amount` (so it is also identical), reuse the serialization function, and fix bug in legacy.

If the serialization is *different*, please fix the bugs in core/ and python/, and keep the serialization function separate.

**Comment by matejcik** [comment]
also leaving a note for myself to rename "manage offer op" to "manage sell offer op" later

**Comment by MisterTicot** [comment]
> i'm just now noticing that your code in core serializes **selling** then **buying** asset, but the legacy code does the opposite, i.e. **buying** then **selling**. which is correct?
> 
> i assumed that the code in core is correct, and the serialization for buy-offer and sell-offer is identical.

Good catch, thanks!

Indeed, [both ops are serialized the same way](https://github.com/stellar/js-stellar-base/blob/master/xdr/Stellar-transaction.x#L121-L149).

**Comment by MisterTicot** [comment]
@matejcik 

Just re-read the code... Where did I serialize the wrong way? Couldn't find it.

**Comment by matejcik** [comment]
https://github.com/trezor/trezor-firmware/pull/627/files#diff-771a1c4ded40f4b313fffe1c5aa274d6R554-R558

**Comment by MisterTicot** [comment]
@matejcik Thanks, I'm fixing it. I'm struggling on something: what's the right way to handle typecheck here?

> def write_manage_offer_op(w, msg: StellarManageOfferOp):

**Comment by matejcik** [comment]
> @matejcik Thanks, I'm fixing it. I'm struggling on something: what's the right way to handle typecheck here?
> 
> > def write_manage_offer_op(w, msg: StellarManageOfferOp):

`Union[StellarManageOfferOp, StellarManageBuyOfferOP]`

**Comment by matejcik** [comment]
> > def write_manage_offer_op(w, msg: StellarManageOfferOp):

that `w` is untyped here tells me that this code is not typechecked yet, so don't worry about it too much. Properly typing the hashwriters is an ongoing task anyway.

**Comment by MisterTicot** [comment]
@matejcik Is there a command that builds and run all tests at once?

**Comment by matejcik** [comment]
unfortunately not. relevant for this PR is:
* in core, `make test` and `make test_emu`
* in legacy, `./script/test` (after you built the legacy emulator)

**Comment by MisterTicot** [comment]
@matejcik Fixed & rebased - but got a problem with `Union`.

**Comment by matejcik** [comment]
the symbol must be [imported in a false branch](https://github.com/trezor/trezor-firmware/blob/bb2556a22cd3ffdc612a06da7d40eeb36a38de10/core/src/trezor/ui/checklist.py#L7)

(to the list of tests above, i'm adding `make style_check`, `make defs_check` and `make gen_check` from the repository root. That is what Travis is running.)

**Comment by matejcik** [comment]
T1 build is failing: https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/381773764

you can build all firmware variants with Docker and see if there is a problem by running `./build-docker.sh`, or refer to https://github.com/trezor/trezor-firmware/blob/master/docs/legacy/index.md

**Comment by MisterTicot** [comment]
(not all the commands you mentioned works here - so I did my best with it)

File: core/src/apps/stellar/operations/layout.py [hunk 5]
```diff
@@ -86,6 +92,16 @@
...
+    await _confirm_offer(ctx, text, op)
+
+
+async def confirm_manage_buy_offer_op(ctx, op: StellarManageBuyOfferOp):
+    text = confirm_manage_offer_common(op)
+    await _confirm_buy_offer(ctx, text, op)
+
+
+def confirm_manage_offer_common(op: StellarManageOfferUnion):
     if op.offer_id == 0:
```
Comment (matejcik) [inline core/src/apps/stellar/operations/layout.py]: The test is passing in an operation where `offer_id` is `None`, which goes to the else-branch here, and then fails to format as number.

Given that `write_manage_offer_op` is serializing it as uint64 always, I think it's a problem with the test, and you should set the offer_id to 0 explicitly?

File: legacy/firmware/stellar.c [hunk 6]
```diff
@@ -481,6 +481,92 @@ bool stellar_confirmManageOfferOp(const StellarManageOfferOp *msg) {
...
+  if (!stellar_signing) return false;
+
+  if (!stellar_confirmSourceAccount(msg->has_source_account,
+                                    msg->source_account)) {
+    stellar_signingAbort(_("Source account error"));
+    return false;
+  }
+
+  // Hash: operation type
+  stellar_hashupdate_uint32(3);
```
Comment (matejcik) [inline legacy/firmware/stellar.c]: this should be the BuyOffer operation type, 3 seems to be SellOffer

File: tests/device_tests/test_msg_stellar_sign_transaction.py [hunk 7]
```diff
@@ -318,3 +318,22 @@ def test_manage_data(client):
...
+    )
+    op.buying_price = 10
+
+    tx = _create_msg()
+
+    response = stellar.sign_tx(client, tx, [op], ADDRESS_N, NETWORK_PASSPHRASE)
+
+    assert (
+        b64encode(response.signature)
+        == b"JBd7YKhZYqRho0nvO6w+r0XD8RiJsNQXrQgWjmo8NimxaEblpOloAHTTA1JMATh0kbhBxbyRnJ1yLAfdzrSBAw=="
```
Comment (matejcik) [inline tests/device_tests/test_msg_stellar_sign_transaction.py]: test on legacy is failing here: https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/384534467
part of it is the operation type, but I fixed that locally and the signature is still different. Leaving this for you to investigate because I don't know what should be correct.

**Comment by matejcik** [comment]
you will also need to register the operation code for BuyOffer in `op_codes` and `op_wire_types` in `core/src/apps/stellar/consts.py`

File: core/src/apps/stellar/operations/layout.py (hunk 5 above)
Comment (MisterTicot) [reply core/src/apps/stellar/operations/layout.py]: I'm not sure about it − that's the original code here − but as far as I know, in transaction XDR `offer_id` is indeed set to 0 when creating new offers.

File: legacy/firmware/stellar.c (hunk 6 above)
Comment (MisterTicot) [reply legacy/firmware/stellar.c]: Got it, thanks

File: core/src/apps/stellar/operations/layout.py (hunk 5 above)
Comment (matejcik) [reply core/src/apps/stellar/operations/layout.py]: what does your wallet do about `offer_id`?

**Comment by MisterTicot** [comment]
That would be nice to have a document listing all the things that need to be done - there's so much places to edit & so much test to run that the when I'll go through adding `pathPaymentStrictSend` I'll remember none of those.

File: core/src/apps/stellar/operations/layout.py (hunk 5 above)
Comment (MisterTicot) [reply core/src/apps/stellar/operations/layout.py]: My wallet, nothing, but I believe `stellar-base` does set it to 0 when the parameter is not provided.

Comment (matejcik) [reply core/src/apps/stellar/operations/layout.py]: makes sense. please modify the test case to set `offer_id` to 0 then

File: tests/device_tests/test_msg_stellar_sign_transaction.py (hunk 7 above)
Comment (MisterTicot) [reply tests/device_tests/test_msg_stellar_sign_transaction.py]: Possibly that `offer_id = 0` thing.

File: core/src/apps/stellar/consts.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
     "StellarChangeTrustOp": const(6),
     "StellarCreateAccountOp": const(0),
     "StellarCreatePassiveOfferOp": const(4),
+    "StellarManageBuyOffer": const(12),
```
Comment (matejcik) [inline core/src/apps/stellar/consts.py]: typo: `ManageBuyOffer`**Op**

File: tests/device_tests/test_msg_stellar_sign_transaction.py (hunk 7 above)
Comment (matejcik) [reply tests/device_tests/test_msg_stellar_sign_transaction.py]: no, the test is still failing with your current changes.
it seems that `price_n` and `price_d` are missing.
AFAICT the C code does not check for presence of fields and uses 0s everywhere, but Python will crash on it.

Comment (MisterTicot) [reply tests/device_tests/test_msg_stellar_sign_transaction.py]: correct

**Comment by MisterTicot** [comment]
Let me know if tests pass, I'll merge commits together to prepare for merging.

**Comment by matejcik** [comment]
tests are still failing, now both core and legacy produces the same signature `IhKLu9XhrrydOlb9U9oBztVO1HMBf1vLi/oc4Okr4gxLDM7oTXzuOccBKfQtxxPv6u4mLtKoFOcjuhHOnXh+AQ==`, which is different from what you are asserting.

I seem to be unable to regenerate the transaction in the Stellar TX builder (the signature I get is also completely different), so I'm afraid this is up to you. Let me know if you encounter problems when running the test; the simplest way to run it without the whole test suite is `pytest tests/device_tests/test_msg_stellar_sign_transaction.py -k buy_offer` while an emulator is running

**Comment by MisterTicot** [comment]
I've tried it, but it stays stuck on:

> collected 11 items / 10 deselected / 1 selected                                                                            
>
> tests/device_tests/test_msg_stellar_sign_transaction.py

I fear I already allocated more time than reasonable on that PR & I have no idea what's wrong. There are other matters that need my attention so I'll have to leave it that way for now.

I think your tool suite is great & has a lot of potential, but contributors efficiency would be greatly improved if it where possible run all the important tasks within a few commands.

It was frustrating to have to deal with dozens of commands & to regularly bump into cryptic error messages that required significant amount of time to get solved. I think I spent more time struggling with the dev environment than working on the code itself.

**Comment by matejcik** [comment]
That's unfortunate. Thanks for your work anyway, hopefully someone can pick this up in the future.

We have #769 about running everything with a single command, so that is something that might happen.

**Comment by MisterTicot** [comment]
That would be immensely helpful.
Let me know, and I'll give it another try once my new system is up & running.

**Comment by overcat** [comment]
> That's unfortunate. Thanks for your work anyway, hopefully someone can pick this up in the future.

Hello @matejcik, can I try to promote this PR?

**Comment by matejcik** [comment]
@overcat feel free to go ahead

although with the code drift, it might be better to start from scratch.

regarding protobuf design: i'm not sure if it's worth it to introduce a separate `ManageBuyOfferOp`. ISTM a better choice is add a `BUY/SELL` enum, which defaults to `SELL` for backwards compatibility.
The shape of the protobuf doesn't match the shape of Stellar XDR anyway, so this shouldn't matter _that much_ -- but what's your opinion?

Also to be clear, we would currently accept the added `ManageBuyOffer`, but not other operations that are missing. (reason basically being that this is "pre-approved" because of this pre-existing PR, but we want to wait for the community firmware for the other additions)

**Comment by overcat** [comment]
Hi @matejcik,

The following comments do not consider compatibility issues.

At least for now, the names of the operations in the protobuf file are the same as those in XDR, I don't think we need to break this rule for this special case. If necessary, we can also share code in a lower level of code. 
The implementation of various software in the current Stellar ecosystem also follows this point. When other Stellar developers try to contribute code to the firmware, I think keeping the same can reduce their confusion.

The above are my opinions, but I am willing to follow your guidelines.

**Comment by matejcik** [comment]
Very well, you're right that keeping it a separate type is better for understandability of outside contributors. The cost of doing it that way is not significant enough to introduce confusion.

**Comment by overcat** [comment]
Hi @matejcik,

So I will submit a new PR, rename `StellarManageOfferOp` to `StellarManageOfferOp` and add a new operation `StellarManageBuyOfferOp`, sounds great?

Also, `StellarPathPaymentOp` has been renamed to `StellarPathPaymentStrictReceiveOp` in XDR and a new operation `StellarPathPaymentStrictSendOp` has been added, of course, I will not add support for `StellarPathPaymentStrictSendOp` before I get permission, but would you like me to rename `StellarPathPaymentOp` to `StellarPathPaymentStrictReceiveOp` in this PR? This might avoid another broken update in the future, so please let me know if I need to do that, thanks.

P.S.

```
struct PathPaymentStrictReceiveOp
{
    Asset sendAsset; // asset we pay with
    int64 sendMax;   // the maximum amount of sendAsset to
                     // send (excluding fees).
                     // The operation will fail if can't be met

    MuxedAccount destination; // recipient of the payment
    Asset destAsset;          // what they end up with
    int64 destAmount;         // amount they end up with

    Asset path<5>; // additional hops it must go through to get there
};
```

```
struct PathPaymentStrictSendOp
{
    Asset sendAsset;  // asset we pay with
    int64 sendAmount; // amount of sendAsset to send (excluding fees)

    MuxedAccount destination; // recipient of the payment
    Asset destAsset;          // what they end up with
    int64 destMin;            // the minimum amount of dest asset to
                              // be received
                              // The operation will fail if it can't be met

    Asset path<5>; // additional hops it must go through to get there
};
```

**Comment by matejcik** [comment]
> So I will submit a new PR, rename `StellarManageOfferOp` to `StellarManageOfferOp` and add a new operation `StellarManageBuyOfferOp`, sounds great?

i think you mean `StellarManageSellOfferOp`, and yes :+1: 

> I will not add support for `StellarPathPaymentStrictSendOp` before I get permission, but would you like me to rename `StellarPathPaymentOp` to `StellarPathPaymentStrictReceiveOp` in this PR

Yes please. Also feel free to implement `StellarPathPaymentStrictSendOp` as well, it really seems symmetrical to the `ManageBuyOffer`/`ManageSellOffer` thing so I believe we can subsume it in here.

**Comment by matejcik** [comment]
@MisterTicot thanks again for your work.

this PR is now superseded by #1838
